### PR TITLE
[API] Change approach to applies-to checking

### PR DESCRIPTION
**Pull request by VWoeltjen**
Some open-ish items on this one, don't know if it is worth addressing these today:

* Naming; is "test" really sufficient
* Want to refactor todo.js a bit more.

w.r.t. the second bullet, I'm thinking of doing something similar to what we do for the Telemetry Adapter tutorial, and treat the to-do view components as something we just have on hand and integrate with the software (to separate out the Zepto code, for instance)

mct.View is of interest here. In particular, handling view life cycle this way simplifies adding metadata, etc. (View is more like a factory and only needs to be instantiated once); since default implementation doesn't do anything meaningful, mct.View is really more of an interface (and could just be documented as such)

### CODE REVIEW

**Comment by VWoeltjen** [comment]
Jotting down some notes on where I'm thinking of going with this (both for myself and reviewers):

* View/ViewDefinition/WhateverWeCallIt
  * Takes a Region instead of an HTML element in "show"; is responsible for populating that Region
    * This also gives us more control over how we build a DOM (most of the time, a Region would be a single element - but it could be the space between some elements, or something else)
    * Also simplifies testing a little (only have to mock a Region with a narrow API, instead of a DOM element)
  * Could also take Region in `canPopulate`; this gives consistent arguments between `canPopulate` and `populate`, and simplifies registration (`mct.view(viewDef)`). This might not be the best thing performance-wise, though.
  * Does not need to expose a `destroy` method or return an analogous function; instead, it observes the region and cleans up after itself on an appropriate event
* Region
  * Has a `show` method accepting DOM elements, fills itself in with those
  * Is an EventEmitter; emits a `clear` event when its contents clear (which Views can listen for to clean up after themselves)
    * An alternative to this would be to take a second argument to `show`, which would be the callback to fire on destroy, to release resources. This argument could be optional (pros and cons to that.) Having it here increases the visibility of the need to handle this (we could log warnings if it was omitted, for instance) and obviates the need for additional interfaces or returned functions from ViewDef. Also means that Region does not have to be an EventEmitter (at the cost of replicating some EventEmitter-esque functionality)
  * Should have some way of checking what kind of region it is ("is this a main region?" would be true in Browse and Edit modes, for instance). Currently using `instanceof` but I consider that harmful. Having a method for this would provide more flexibility (`if (region.isLike(mct.regions.main)) someStuff();`)
  * All of the above loads some functionality into Region that is currently separate. I don't think it needs to be separate and that Region is a coherent thing (an area in the display that stuff can be put in.) "What kind of region is this?" and "Fill that region up with something!" are both sensible things to say about a region, so it passes a horse-sense cohesion test

**Comment by larkin** [comment]
A lot of concerns getting mashed together here and I think we're going backwards. I also think we're overloading the term Region and it would help to clearly split the concepts:

***First***, we have a `Region`: a section of the interface in which views can be shown.  Handles lifecycle of views displayed in that region, but does not choose which view to display in a given region at a give time.

***Secondly***, we have some sort of region mapping or routing: which views are displayed in which area of the screen in response to certain user inputs e.g. navigation, editing, inspection, selection.  E.g. when I navigate to an object what view do I see in the main screen?

The first concept (Regions) is a purely internal construct, an implementation detail, and should be hidden from the public API, but we should look at existing code in designing it. [1]

The second topic is a core concept of our public API that we have called "regions" but should probably use a different name for to reflect how it maps to user interactions / inputs. specific user interactions cause certain views to be displayed in certain regions.  

# tl;dr Moving forward 

We should separate "this is a view of an object" and "this is a view for a certain region of the screen based on a navigation event, selection, or other user input as standardized by Open MCT".

A view will always have to say "yes, I can apply to this domain object", so let's bake that into the API without bringing in any concept of regions, and get that out the door.  Separately, let's get a reasonable placeholder up for view registration so that we can iterate on that while still shipping code.

Now, to divide and conquer: "View" and "view mapping"

## View: Views apply to objects

A view is a way of seeing a object.  They have no concept of regions, they define lifecycle methods as necessary but have no idea how to trigger those lifecycle methods[1].

* `ViewDefinition`: has metadata about what kind of view this is (`name`, `glyph`, etc), and what kinds of objects it can apply to (`test`).  Also defines a factory for instantiating `ViewInstance`s, which takes a `domainObject` and a `container` element.  ***Need to define interface and move forward.***
* `ViewInstance`: an instance of a view, defines how to render, how to be destroyed, can implement other lifecycle methods but does not know when to trigger them.  ***Need to define interface and move forward.***

I have proposed one combined interface for ViewDefinition and ViewInstance [here](https://github.com/nasa/openmct/commit/f68ce2fc80878d90c68a8aff681923c7c989f692#diff-f853b17865fdfd8c86279a146bcdd808R28).  I will take Victor's code here and present a side by side comparison of the methods.

## View Mapping: map user interactions to views

An intermediate API as we figure out what it truly means, we have some future concepts (display component in modal dialog) that can benefit from this.  This seems to be what we have today, and would work for now.

* `mct.views.register(MCT.regions.Main, MyViewDefinition)`.  A view is registered as valid for a specific region-- that is, a area of the screen (main, toolbar, inspector) for a given navigation state (edit, browse, selection).

# Footnotes

[1] Some existing design to look at when considering regions: 
* Derick Bailey on Regions/Region Managers in Backbone: https://lostechies.com/derickbailey/2011/12/12/composite-js-apps-regions-and-region-managers/
* ReactDOM.render, used for bootstrapping react views, e.g. the code that react plugin developers will write to integrate with our APIs: https://facebook.github.io/react/blog/2015/10/01/react-render-and-top-level-api.html

[2] lifecycle methods are optional, implemented on the view.  The region or whatever is displaying the view (perhaps a layout) determines when to trigger them.  Additional methods we may want are things such as "onResize" (my container has resized so I should reflow/recalculate styles), and "onShow" (I'd like to fade in, please).  Marionette [regions](http://marionettejs.com/docs/v2.4.7/marionette.region.html) and [views](http://marionettejs.com/docs/v2.4.7/marionette.view.html) define a number of lifecycle events we may find useful, as do [React components](https://facebook.github.io/react/docs/component-specs.html)

**Comment by VWoeltjen** [comment]
Tweaked region to match Backbone.Marionette's version a little more closely (also able to remove this from public API as a consequence.)

Split out View and ViewDefinition as described in "Views apply to objects" above. Electing to use two separate interfaces instead of using static properties on constructors; that would reduce number of interfaces while increasing their complexity, and reducing flexibility somewhat. Mostly I was worried about how to document that; I've rarely seen function/constructor with properties show up in APIs/SPIs, and when I do see them they've been a source of some confusion (e.g. Angular's `$inject` property)

View and ViewDefinition both end up being really-just-interfaces in this case. I think that's a good thing (no need to inherit from a base class, easy to substitute) but it's worth pointing out. I left the JS files in place for sake of JSDoc and review clarity.

Lastly, switched view registration back to the `mct.view(mct.regions.main, ...)` as suggested; it's completely true that this works for now, and we don't really gain anything by cramming it into the `canView` check.

**Comment by larkin** [comment]
@VWoeltjen thumbs up here, this is simple and straightforward.  I like keeping the JS files around for JSDoc.

If it's cool with you I'd like to merge this into api-tutorials.

**Comment by VWoeltjen** [comment]
> If it's cool with you I'd like to merge this into api-tutorials.

Go for it! No more changes I'd like to make
